Disable broken tests; force CI to run on every PR

cjbarth opened the pull request:
To improve code quality going forward, this PR will enable all tests to run on every PR. To get this to run the first time, broken tests have been disabled.

LoneRifle commented:
Could you clarify what you mean by broken tests? I just did `npm test` on my local dev and on a gitpod instance using Node 16 as well as Node 4, and all the tests pass.

cjbarth commented:
This is why we have code review :)

It turns out that `nodeunit` doesn't do globbing very well. Once I reverted the glob, all the tests started working. Sorry about that.

So, the only thing that this PR would do then is to make sure all tests run on every PR.

LoneRifle commented on .github/workflows/ci.yml:
```diff
@@ -1,19 +1,33 @@
-          node-version: ${{ matrix.version }}
-      - run: npm install
+          node-version: ${{ matrix.node-version }}
+      - run: npm ci
       - run: npm test
+      - run: npm update
+      - run: npm ci
+      - run: npm test
```
Couple of things:
- Why the need to run `npm ci; npm test` twice? are we trying to test the codebase against the deps defined in package-lock as well as the deps after updating said lock?
- You may consider the use of `if` to [conditionally](https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution) run commands, so that for newer node versions we run `npm ci`, and for older ones we run `npm install`.
  - This is of course dependent on whether we want to continue supporting node 4 and 6. Limiting ourselves to node 8 would give us the freedom to rewrite the codebase with more modern language constructs (like async/await and `const` and `let` instead of `var`), but that depends on whether we get round to doing so.

cjbarth replied:
I'm acting as if we are going to do a `semver-major` release to drop older versions of Node. I really don't see a need to keep supporting something that old.


> are we trying to test the codebase against the deps defined in package-lock as well as the deps after updating said lock?

Yes, that is the idea. We had some problems with `node-saml` and `passport-saml` where people would report bugs that we couldn't reproduce. We found it was actually due to an incompatibility with a dependency. We actually found a problem with a TypeScript incompatibility/regression. Maybe I can mark these as `experimental` so that we don't fail the entire PR on them? I just don't want problem like that to remain hidden for very long, so I hesitate to do even that.

LoneRifle commented on .github/workflows/ci.yml:
```diff
@@ -1,19 +1,33 @@
     strategy:
       matrix:
-        version: [8, lts/*, latest]
+        node-version: [8.x, 10.x, 12.x, 14.x, 16.x, 18.x]
```
- for conciseness, it should suffice to specify just `n` instead of `n.x`
- I would like to retain the `lts/*` and `latest` aliases, so that in the event that no one is available to update the list of versions, we have some assurance that we are testing against actively maintained versions of node

cjbarth replied:
For reference, I think that this might be good to keep in mind: https://github.com/actions/setup-node/issues/26#issuecomment-1148828915

LoneRifle reviewed (approved):
Generally lgtm. Some concerns about the nature of the changes but non-blockers.

shunkica commented:
> It turns out that `nodeunit` doesn't do globbing very well. Once I reverted the glob, all the tests started working. Sorry about that.

Maybe this is an opportunity to move away from `nodeunit` as it is a dead project.

cjbarth commented:
@shunkica , I'd love to see a PR that does that. We currently use [Jasmine](https://jasmine.github.io/) in the other projects here, so it would be reasonable to move to that for consistency. However, anything would be better than `nodeunit` at this point.

cjbarth commented:
@shunkica , would you be willing to put up a PR for a migration of the tests?

cjbarth commented:
@LoneRifle , I think I've addressed all your concerns. Feel free to have another look at this. Your feedback is quite good. I look forward implementing some of this over in `node-saml` and `passport-saml`.

LoneRifle reviewed (approved):
lgtm, as discussed. Thanks for looking into this!
